simulator/risc_v/riscv_decoder.h

```diff
@@ -58,45 +105,177 @@ struct RISCVInstrDecoder
+        return ( apply_mask( CL_imm_1, 0b111) << 3U)
+             | ( apply_mask( CL_imm_2, 0b11) << 6U);
+    }
+
+    constexpr uint32 get_C_ADDI_immediate() const noexcept
+    {
+        return ( apply_mask( CI_imm_1, 0b1) << 5U)
+             | ( apply_mask( CI_imm_2, 0b11111));
+    }
+
+    constexpr uint32 get_C_J_immediate() const noexcept
+    {
+        return ( apply_mask( CJ_imm, 0b10000000000) << 11U)
+             | ( apply_mask( CJ_imm, 0b01000000000) << 4U)
+             | ( apply_mask( CJ_imm, 0b00110000000) << 8U)
+             | ( apply_mask( CJ_imm, 0b00001000000) << 10U)
+             | ( apply_mask( CJ_imm, 0b00000100000) << 6U)
+             | ( apply_mask( CJ_imm, 0b00000010000) << 7U)
+             | ( apply_mask( CJ_imm, 0b00000001110) << 1U)
+             | ( apply_mask( CJ_imm, 0b00000000001) << 5U);
+    }
+
+    constexpr uint32 get_C_LUI_immediate() const noexcept
+    {
+        return ( apply_mask( CI_imm_1, 0b1) * 0xfffe0)
+             | ( apply_mask( CI_imm_2, 0b11111));
+    }
+
+    constexpr uint32 get_C_SRLI_immediate() const noexcept
+    {
+        return ( apply_mask( CI_imm_1, 0b1) << 5U)
+             | ( apply_mask( CI_imm_2, 0b11111));
+    }
+
+    constexpr uint32 get_C_BEQZ_immediate() const noexcept
+    {
+        return ( apply_mask( CL_imm_1, 0b100) << 8U)
+             | ( apply_mask( CL_imm_1, 0b011) << 3U)
+             | ( apply_mask( CI_imm_2, 0b11000) << 6U)
+             | ( apply_mask( CI_imm_2, 0b00110) << 1U)
+             | ( apply_mask( CI_imm_2, 0b00001) << 5U);
+    }
+
+    constexpr uint32 get_C_ADDI4SPN_immediate() const noexcept
+    {
+        return ( apply_mask( CIW_imm, 0b11000000) << 4U)
+             | ( apply_mask( CIW_imm, 0b00111100) << 6U)
+             | ( apply_mask( CIW_imm, 0b00000010) << 2U)
+             | ( apply_mask( CIW_imm, 0b00000001) << 3U);
+    }
+
+    constexpr uint32 get_C_ADDI16SP_immediate() const noexcept
+    {
+        return ( apply_mask( CI_imm_1, 0b1) << 9U)
+             | ( apply_mask( CI_imm_2, 0b10000) << 4U)
+             | ( apply_mask( CI_imm_2, 0b01000) << 6U)
+             | ( apply_mask( CI_imm_2, 0b00110) << 7U)
+             | ( apply_mask( CI_imm_2, 0b00001) << 5U);
+    }
+
     uint32 get_immediate_value( char subset) const noexcept
     {
         switch (subset) {
-        case 'I': return I_imm;
-        case 'B': return get_B_immediate();
-        case 'S': return S_imm4_0 | (S_imm11_5 << 5);
-        case 'U': return U_imm;
-        case 'J': return get_J_immediate();
-        case 'C': return csr_imm;
-        case ' ': return 0;
-        default: assert(0); return 0;
+        case 'I':    return I_imm;
+        case 'B':    return get_B_immediate();
+        case 'S':    return S_imm4_0 | (S_imm11_5 << 5U);
+        case 'U':    return U_imm;
+        case 'J':    return get_J_immediate();
+        case 'C':    return csr_imm;
+        case ' ':    return 0;
+        case C_LWSP: return get_C_LWSP_immediate();
```

Comment:
Can this method be separated here into two methods (`get_immediate_value` and `get_compressed_immediate_value`)?